File: src/main/java/guru/xgm/jass/openapi/actionSystem/convert/Jass2AnyVisitor.java
package guru.xgm.jass.openapi.actionSystem.convert;

import com.intellij.psi.PsiElement;
import guru.xgm.jass.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public abstract class Jass2AnyVisitor extends JassVisitor {

    public StringBuffer stringBuffer = new StringBuffer();

    public void appendSingleLineComment(String comment) {
        stringBuffer.append("//").append(comment).append("\n");
    }

    public void appendStatementLineEnd() {
        stringBuffer.append(";\n");
    }

    public void appendString(@NotNull String str) {
        stringBuffer.append(str);
    }

    public void appendReal(@NotNull String real) {
        stringBuffer.append(real);
    }

    public void appendNull() {
        stringBuffer.append("null");
    }

    public void appendFunctionAsCode(@NotNull String name) {
        stringBuffer.append(name);
    }

    public @NotNull String getConvertedTypeName(@NotNull String type) {
        return type;
    }

    // --- globals
    @Override
    public void visitGlob(@NotNull JassGlob o) {
        appendSingleLineComment("globals");
        o.acceptChildren(this);
        appendSingleLineComment("endglobals");
    }

    // --- variable
    public abstract void appendVar(boolean constant, boolean global, boolean array, @NotNull String type, String name, @Nullable JassExpr expr);

    private void appendVarPrepare(boolean constant, boolean global, @NotNull JassVar var) {
        appendVar(constant, global, var.getArray() != null, getConvertedTypeName(var.getTypeName().getText()), var.getId().getText(), var.getExpr());
    }

    @Override
    public void visitGvar(@NotNull JassGvar o) {
        appendVarPrepare(o.getConstant() != null, true, o.getVar());
    }

    @Override
    public void visitLvarStmt(@NotNull JassLvarStmt o) {
        appendVarPrepare(false, false, o.getVar());
    }

    // --- function
    public abstract void appendFunction(@Nullable String returns, String name, @NotNull List<JassParam> params, @NotNull List<JassStmt> statements);

    @Override
    public void visitFun(@NotNull JassFun o) {
        final JassFunRet ret = o.getFunRet();
        final @Nullable var rettype = ret != null ? ret.getTypeName() : null;

        final var name = o.getId();

        final var take = o.getFunTake();
        List<JassParam> params = new ArrayList<>();
        if (take != null) {
            final var list = take.getParamList();
            if (list != null) params = list.getParamList();
        }

        appendFunction(
                ret == null || rettype == null || ret.getNothing() != null ? null : getConvertedTypeName(rettype.getText()),
                name == null ? "" : name.getText(),
                params,
                o.getStmtList()
        );
    }

    // --- statement

    // return
    @Override
    public void visitReturnStmt(@NotNull JassReturnStmt o) {
        stringBuffer.append("return ");
        final var expr = o.getExpr();
        if (expr != null) expr.accept(this);
        appendStatementLineEnd();
    }

    // set
    @Override
    public void visitSetStmt(@NotNull JassSetStmt o) {
        System.out.print("lua\n" + o.getText());

        final var id = o.getId();
        if (id != null) stringBuffer.append(id.getText());
        final var arr = o.getArrayAccess();
        if (arr != null) arr.accept(this);
        final var expr = o.getExpr();
        if (expr != null) {
            stringBuffer.append(" = ");
            expr.accept(this);
        }
        appendStatementLineEnd();
    }

    // ===============

    private void unexpr(@Nullable JassExpr expr, String op) {
        if (expr == null) return;
        stringBuffer.append(" ").append(op).append(" ");
        expr.accept(this);
    }

    private boolean text(@Nullable PsiElement pe) {
        if (pe == null) return false;
        stringBuffer.append(pe.getText());
        return true;
    }

    @Override
    public void visitExitWhenStmt(@NotNull JassExitWhenStmt o) {
        final var expr = o.getExpr();
        if (expr == null) return;
        stringBuffer.append("if (");
        expr.accept(this);
        stringBuffer.append(") break;\n");
    }

    @Override
    public void visitLoopStmt(@NotNull JassLoopStmt o) {
        stringBuffer.append("while (true) {\n");
        for (JassStmt stmt : o.getStmtList()) stmt.accept(this);
        stringBuffer.append("}\n");
    }

    @Override
    public void visitCallStmt(@NotNull JassCallStmt o) {
        if (o.getDebug() != null) stringBuffer.append("/*debug*/");
        o.getFunCall().accept(this);
        appendStatementLineEnd();
    }

    @Override
    public void visitStmt(@NotNull JassStmt o) {
        o.acceptChildren(this);
    }

    @Override
    public void visitArgList(@NotNull JassArgList o) {
        final var list = o.getExprList();
        for (int i = 0; i < list.size(); i++) {
            list.get(i).accept(this);
            if (i < list.size() - 1) stringBuffer.append(", ");
        }
    }

    @Override
    public void visitFunCall(@NotNull JassFunCall o) {
        stringBuffer.append(o.getId().getText()).append("(");

        final var list = o.getArgList();
        if (list != null) list.accept(this);

        stringBuffer.append(")");
    }

    @Override
    public void visitArrayAccess(@NotNull JassArrayAccess o) {
        stringBuffer.append(o.getId().getText()).append("[");
        final var expr = o.getExpr();
        if (expr != null) expr.accept(this);
        stringBuffer.append("]");
    }

    @Override
    public void visitFuncAsCode(@NotNull JassFuncAsCode o) {
        appendFunctionAsCode(o.getId().getText());
    }

    @Override
    public void visitPrimExpr(@NotNull JassPrimExpr o) {
        if (text(o.getTrue())) return;
        if (text(o.getFalse())) return;
        final var tnull = o.getNull();
        if (tnull != null) {
            appendNull();
            return;
        }
        if (text(o.getId())) return;
        if (text(o.getIntval())) return;
        final var str = o.getStrval();
        if (str != null) {
            appendString(str.getText());
            return;
        }
        final var raw = o.getRawval();
        if (raw != null) {
            stringBuffer.append("FourCC(\"").append(raw.getText().replace("'", "")).append("\")");
            return;
        }
        final var hex = o.getHexval();
        if (hex != null) {
            stringBuffer.append(hex.getText().replace("$", "0x"));
            return;
        }
        final var real = o.getRealval();
        if (real != null) {
            appendReal(real.getText());
            return;
        }
        o.acceptChildren(this);
    }

    // --- expression
    public void acceptExpr(@Nullable JassExpr expr) {
        if (expr == null) return;
        expr.accept(this);
    }

    public void exprListConcatByOperator(List<JassExpr> list, String op) {
        list.get(0).accept(this);
        stringBuffer.append(" ").append(op).append(" ");
        list.get(1).accept(this);
    }

    @Override
    public void visitMulUnExpr(@NotNull JassMulUnExpr o) {
        unexpr(o.getExpr(), "*");
    }

    @Override
    public void visitDivUnExpr(@NotNull JassDivUnExpr o) {
        unexpr(o.getExpr(), "/");
    }

    @Override
    public void visitPlusUnExpr(@NotNull JassPlusUnExpr o) {
        unexpr(o.getExpr(), "+");
    }

    @Override
    public void visitMinusUnExpr(@NotNull JassMinusUnExpr o) {
        unexpr(o.getExpr(), "-");
    }

    @Override
    public void visitNotExpr(@NotNull JassNotExpr o) {
        unexpr(o.getExpr(), "!");
    }

    @Override
    public void visitEqExpr(@NotNull JassEqExpr o) {
        exprListConcatByOperator(o.getExprList(), "==");
    }

    @Override
    public void visitNeqExpr(@NotNull JassNeqExpr o) {
        exprListConcatByOperator(o.getExprList(), "!=");
    }

    @Override
    public void visitLtExpr(@NotNull JassLtExpr o) {
        exprListConcatByOperator(o.getExprList(), "<");
    }

    @Override
    public void visitLtEqExpr(@NotNull JassLtEqExpr o) {
        exprListConcatByOperator(o.getExprList(), "<=");
    }

    @Override
    public void visitGtExpr(@NotNull JassGtExpr o) {
        exprListConcatByOperator(o.getExprList(), ">");
    }

    @Override
    public void visitGtEqExpr(@NotNull JassGtEqExpr o) {
        exprListConcatByOperator(o.getExprList(), ">=");
    }

    @Override
    public void visitParenExpr(@NotNull JassParenExpr o) {
        stringBuffer.append("(");
        acceptExpr(o.getExpr());
        stringBuffer.append(")");
    }

    @Override
    public void visitAndExpr(@NotNull JassAndExpr o) {
        exprListConcatByOperator(o.getExprList(), "&&");
    }

    @Override
    public void visitOrExpr(@NotNull JassOrExpr o) {
        exprListConcatByOperator(o.getExprList(), "||");
    }

    @Override
    public void visitPlusExpr(@NotNull JassPlusExpr o) {
        exprListConcatByOperator(o.getExprList(), "+");
    }

    @Override
    public void visitMinusExpr(@NotNull JassMinusExpr o) {
        exprListConcatByOperator(o.getExprList(), "-");
    }

    @Override
    public void visitMulExpr(@NotNull JassMulExpr o) {
        exprListConcatByOperator(o.getExprList(), "*");
    }

    @Override
    public void visitDivExpr(@NotNull JassDivExpr o) {
        exprListConcatByOperator(o.getExprList(), "/");
    }


}

File: src/main/kotlin/raft/war/language/jass/action/Jass2AnyVisitor.kt
package raft.war.language.jass.action

import com.intellij.psi.PsiElement
import raft.war.language.jass.inspection.number.JassRawcode
import raft.war.language.jass.psi.*

abstract class Jass2AnyVisitor : JassVisitor() {
    private fun getSafeName(name: String): String {
        if (keywords.contains(name)) return name + KEYWORD_SUFFIX
        return name
    }

    fun appendSafeName(name: String) {
        stringBuffer.append(name)
        if (keywords.contains(name)) stringBuffer.append(KEYWORD_SUFFIX)
    }


    var keywords: HashSet<String> = HashSet()

    var stringBuffer: StringBuffer = StringBuffer()

    open fun appendSingleLineComment(comment: String?) {
        stringBuffer.append("//").append(comment).append("\n")
    }

    open fun appendBlockLineComment(comment: String?) {
        stringBuffer.append("/*").append(comment).append("*/")
    }

    open fun appendStatementLineEnd() {
        stringBuffer.append(";\n")
    }

    private fun appendString(str: String) {
        stringBuffer.append(str)
    }

    open fun appendReal(real: String) {
        stringBuffer.append(real)
    }

    open fun appendNull() {
        stringBuffer.append("null")
    }

    open fun appendFunctionAsCode(name: String) {
        stringBuffer.append(name)
    }

    open fun getConvertedTypeName(type: String): String {
        return type
    }

    // --- typedef
    open fun appendTypeDef(name: String, base: String) {
        appendSingleLineComment("$name extends $base")
    }

    override fun visitTypeDef(o: JassTypeDef) {
        val name = o.typeName
        val base = o.typeExtends?.typeName
        if (name == null || base == null) return
        appendTypeDef(name.text, base.text)
    }

    // --- globals
    override fun visitGlob(o: JassGlob) {
        appendSingleLineComment("globals")
        o.acceptChildren(this)
        appendSingleLineComment("endglobals")
    }

    // --- variable
    abstract fun appendVar(
        constant: Boolean,
        global: Boolean,
        array: Boolean,
        type: String,
        name: String?,
        expr: JassExpr?
    )

    private fun appendVarPrepare(constant: Boolean, global: Boolean, varDef: JassVarDef) {
        appendVar(
            constant,
            global,
            varDef.array != null,
            getConvertedTypeName(varDef.typeName.text),
            getSafeName(varDef.varName.text),
            varDef.expr
        )
    }

    override fun visitGvar(o: JassGvar) = appendVarPrepare(o.constant != null, true, o.getVarDef())

    override fun visitLvarStmt(o: JassLvarStmt) =
        appendVarPrepare(constant = false, global = false, varDef = o.getVarDef())

    // --- function
    abstract fun appendFunction(returns: String?, name: String?, params: List<JassParam?>, statements: List<JassStmt?>)

    override fun visitFun(o: JassFun) {
        val head = o.funHead

        val ret = head.funRet
        val rettype = ret?.typeName

        val name = head.funName

        val take = head.funTake
        var params: List<JassParam?> = ArrayList()
        if (take != null) {
            val list = take.paramList
            if (list != null) params = list.paramList
        }

        appendFunction(
            if (ret == null || rettype == null || ret.nothing != null) null else getConvertedTypeName(rettype.text),
            if (name == null) "" else getSafeName(name.text),
            params,
            o.funBody.stmtList
        )
    }

    // --- statement
    // stmt list
    override fun visitStmt(o: JassStmt) = o.acceptChildren(this)

    // return
    override fun visitReturnStmt(o: JassReturnStmt) {
        stringBuffer.append("return ")
        acceptExpr(o.expr)
        appendStatementLineEnd()
    }

    // set
    override fun visitSetStmt(o: JassSetStmt) {
        val vn = o.varName
        if (vn != null) appendSafeName(vn.text)
        val arr = o.arrayAccess
        arr?.accept(this)
        val expr = o.expr
        if (expr != null) {
            stringBuffer.append(" = ")
            expr.accept(this)
        }
        appendStatementLineEnd()
    }

    // call
    override fun visitCallStmt(o: JassCallStmt) {
        if (o.debug != null) appendBlockLineComment("debug")
        o.funCall?.accept(this)
        appendStatementLineEnd()
    }

    override fun visitFunCall(o: JassFunCall) {
        val name = o.funName.text
        val list = o.argList
        if (name == "ExecuteFunc" && list != null) {
            val exprs = list.exprList
            if (exprs.size == 1) {
                val expr = exprs[0].firstChild
                if (expr.node.elementType === JassTypes.STRVAL) {
                    val fn = expr.text.replaceFirst("^\"".toRegex(), "").replaceFirst("\"$".toRegex(), "")
                    if (fn.matches("^[a-zA-Z_][a-zA-Z_0-9]*$".toRegex())) {
                        stringBuffer.append(fn).append("()")
                        return
                    }
                }
            }
        }

        appendSafeName(name)
        stringBuffer.append("(")
        list?.accept(this)
        stringBuffer.append(")")
    }

    override fun visitArgList(o: JassArgList) {
        val list = o.exprList
        for (i in list.indices) {
            list[i].accept(this)
            if (i < list.size - 1) stringBuffer.append(", ")
        }
    }

    override fun visitFunRef(o: JassFunRef) {
        appendFunctionAsCode(getSafeName(o.funName!!.text))
    }

    // ===============
    private fun text(pe: PsiElement?): Boolean {
        if (pe == null) return false
        stringBuffer.append(pe.text)
        return true
    }

    override fun visitArrayAccess(o: JassArrayAccess) {
        stringBuffer.append(o.varName.text).append("[")
        val expr = o.expr
        expr?.accept(this)
        stringBuffer.append("]")
    }

    // --- expression
    fun acceptExpr(expr: JassExpr?) {
        if (expr == null) return
        expr.accept(this)
    }

    fun appendExprWithPrefixOp(expr: JassExpr?, op: String?) {
        if (expr == null) return
        stringBuffer.append(" ").append(op).append(" ")
        expr.accept(this)
    }

    fun appendExprListConcatByOperator(list: List<JassExpr>, op: String?) {
        list[0].accept(this)
        stringBuffer.append(" ").append(op).append(" ")
        list[1].accept(this)
    }

    open fun appendRawcode(raw: JassRawcode) {
        stringBuffer.append("FourCC(\"").append(raw.strval).append("\")")
    }

    override fun visitNum(o: JassNum) {
        if (text(o.intval)) return

        val raw = o.rawval
        if (raw != null) {
            appendRawcode(JassRawcode(raw))
            return
        }
        val hex = o.hexval
        if (hex != null) {
            stringBuffer.append(hex.text.replace("$", "0x"))
            return
        }
        val real = o.realval
        if (real != null) {
            appendReal(real.text)
            return
        }
    }

    // primary
    override fun visitPrimExpr(o: JassPrimExpr) {
        if (text(o.getTrue())) return
        if (text(o.getFalse())) return
        val tnull = o.getNull()
        if (tnull != null) {
            appendNull()
            return
        }
        val vn = o.varName
        if (vn != null) {
            appendSafeName(vn.text)
            return
        }

        val str = o.str
        if (str != null) {
            appendString(str.text)
            return
        }

        o.acceptChildren(this)
    }

    override fun visitMulUnExpr(o: JassMulUnExpr) = appendExprWithPrefixOp(o.expr, "*")
    override fun visitDivUnExpr(o: JassDivUnExpr) = appendExprWithPrefixOp(o.expr, "/")
    override fun visitPlusUnExpr(o: JassPlusUnExpr) = appendExprWithPrefixOp(o.expr, "+")
    override fun visitMinusUnExpr(o: JassMinusUnExpr) = appendExprWithPrefixOp(o.expr, "-")
    override fun visitNotExpr(o: JassNotExpr) = appendExprWithPrefixOp(o.expr, "!")
    override fun visitEqExpr(o: JassEqExpr) = appendExprListConcatByOperator(o.exprList, "==")
    override fun visitNeqExpr(o: JassNeqExpr) = appendExprListConcatByOperator(o.exprList, "!=")
    override fun visitLtExpr(o: JassLtExpr) = appendExprListConcatByOperator(o.exprList, "<")
    override fun visitLtEqExpr(o: JassLtEqExpr) = appendExprListConcatByOperator(o.exprList, "<=")
    override fun visitGtExpr(o: JassGtExpr) = appendExprListConcatByOperator(o.exprList, ">")
    override fun visitGtEqExpr(o: JassGtEqExpr) = appendExprListConcatByOperator(o.exprList, ">=")
    override fun visitParenExpr(o: JassParenExpr) {
        stringBuffer.append("(")
        acceptExpr(o.expr)
        stringBuffer.append(")")
    }

    override fun visitAndExpr(o: JassAndExpr) = appendExprListConcatByOperator(o.exprList, "&&")
    override fun visitOrExpr(o: JassOrExpr) = appendExprListConcatByOperator(o.exprList, "||")
    override fun visitPlusExpr(o: JassPlusExpr) = appendExprListConcatByOperator(o.exprList, "+")
    override fun visitMinusExpr(o: JassMinusExpr) = appendExprListConcatByOperator(o.exprList, "-")
    override fun visitMulExpr(o: JassMulExpr) = appendExprListConcatByOperator(o.exprList, "*")
    override fun visitDivExpr(o: JassDivExpr) = appendExprListConcatByOperator(o.exprList, "/")

    companion object {
        private const val KEYWORD_SUFFIX = "_FUCKING_KEYWORD"
    }
}
